ENH: distutils: copy msvcp*.dll into extra_dll_dir

xoviat opened the pull request:
Closes scipy/scipy#7969.

Issues with this fix (feel free to add more):

- [x] It doesn't create the `__config__` if there are not fortran objects (the config should be written when there are either fortran OR c++ objects and the compiler is MSVC).

xoviat committed:
ENH: distutils: copy msvcp*.dll into extra_dll_dir

Closes scipy/scipy#7969.

xoviat committed:
BUG: Fix detection of MSVC++ location

Use setuptools to detect the "redist" directory location

pv commented:
I think this is only needed if `need_cxx_compiler`. Note that `extra-dll` is added to PATH in `__config__` unconditionally, if the directory exists, so that should be fine.

xoviat commented:
> I think this is only needed if need_cxx_compiler. 

You're correct but was easier to test because numpy has no CXX.

Unfortunately we cannot help out Python 2.7 or Python 3.4 without magic because MS put installers rather than DLLs in the 'redist' folder.

xoviat commented:
I think that's pretty much it other than maybe cleaning up my horrible commit messages. As I said, it would be very difficult of fix 2.7 or 3.4 unless someone wrote a `find_msvcp` that downloaded and unpacked the redistributables from MS.

xoviat commented:
Before I clean up the commits, someone will need to approve these changes (that doesn't mean that someone should approve these changes; if they don't, I will make more modifications and then cleanup the commits at the end if this PR is approved).

xoviat commented:
Also, should probably wait for the CI before approving.

pv commented on numpy/distutils/command/build_ext.py:
```diff
@@ -566,6 +584,24 @@ def _libs_with_msvc_and_fortran(self, fcompiler, c_libraries,
+    def _get_msvcp_dll(self):
+        assert self.compiler.compiler_type == 'msvc'
+
+        vc_root = self.compiler.cc
```
The compiler object doesn't have the `cc` attribute if all files were already compiled on a previous round.
So you need to add
```
if not self.compiler.initialized:
    self.compiler.initialize()
```
before that.

xoviat replied:
Will do.

pv reviewed (commented):
I guess this looks like the correct place to copy the files over.
Within distutils, I don't see other good places to insert this.

We could also think about doctoring the generated wheels afterwards,
but I think it's simpler if numpy.distutils just does the right thing.

xoviat commented on numpy/distutils/command/build_ext.py:
```diff
@@ -566,6 +584,24 @@ def _libs_with_msvc_and_fortran(self, fcompiler, c_libraries,
+
+        vc_root = self.compiler.cc
+        while os.path.basename(vc_root).lower() != 'vc' and vc_root:
+            vc_root = os.path.dirname(vc_root)
```
The problem with this approach is that AFAIK, `MSVCComplier` does not have a `version` attribute. So I guess we can add the path for MSVC 2015-2017 and hope that it doesn't change again.

xoviat replied:
@pv What do you want to do about this?

pv replied:
I don't care too much either way, I was just wondering if this is more robust than hardcoding in practice.

xoviat replied:
It's definitely more robust than hardcoding, but that doesn't mean that it won't break in the future. I wish that MS would provide an API for this or something.

pv replied:
setuptools.msvc has some more advanced vc stuff. don't know if they guarantee that as api, likely they don't

xoviat replied:
Actually that's a better plan.

xoviat commented on numpy/distutils/command/build_ext.py:
```diff
@@ -271,6 +274,21 @@ def run(self):
+            print('copying ' + msvcp_dll + ' -> ' + dst)
+            with open(msvcp_dll, 'rb') as fsrc:
+                with open(dst, 'wb+') as fdst:
+                    shutil.copyfileobj(fsrc, fdst)
```
Nope. The comment explains it:

># Copy in an odd manner to avoid permission errors

`shutil.copyfile` may use the winapi to copy the file, which may copy the permissions as well. The permissions on that particular file make it read only; the `wheel` plugin cannot move it later when it wants.

xoviat commented on numpy/distutils/command/build_ext.py:
```diff
@@ -566,6 +584,24 @@ def _libs_with_msvc_and_fortran(self, fcompiler, c_libraries,
+
+        return next(iglob(
+            vc_root + '/redist/' + specifier + '/**/msvcp*.dll',
+            recursive=True))
```
Will do.

xoviat commented on numpy/distutils/command/build_ext.py:
```diff
@@ -271,6 +274,21 @@ def run(self):
+        # See scipy/scipy#7969 for more information
+        # Copy in an odd manner to avoid permission errors
+        if self.compiler.compiler_type == 'msvc' and need_cxx_compiler \
+                and sys.version_info >= (3, 5):
```
> At least the "microsoft visual c++ 9.0 for python" package doesn't seem to ship the msvcp90.dll, so maybe we just ignore those releases and just trust they're already installed. 

Exactly.

> If so, add a comment.
 Will do.

xoviat commented on numpy/distutils/command/build_ext.py:
```diff
@@ -23,6 +24,8 @@
 from numpy.distutils.command.config_compiler import show_fortran_compilers
 
+if sys.version_info >= (3, 5):
+    from glob import iglob
```
Will do.

pv commented:
I guess would be useful as a next step to check if it results to OK scipy wheels (don't have time to look at that before tomorrow/utc+2), then do as before (or maybe just merge asap).

xoviat commented:
Give me a second. I'll file a PR against scipy that simply changes the distutils repository.

xoviat commented:
I think we can merge within 20 minutes if you want. I have an appveyor running to check the functionality. 

https://ci.appveyor.com/project/xoviat/scipy/build/job/stmct3mmknnn9sup

xoviat commented:
Assuming [the build](https://ci.appveyor.com/project/xoviat/scipy/build/1.0.4) passes, this should be okay. Obviously we should merge soon so that we can produce scipy wheels with the DLL included.

xoviat commented:
@pv This is the best that I could come up with. I have confirmed that it copies the DLL on Python 3.5 and 3.6.

pv commented:
The setuptools.msvc API is not public, so the code should fail 
gracefully (ie. just show some warning message) if it doesn't contain 
what we expected.

Also, I think it would be better to move the setuptools import to inside 
`_get_msvcp()` (where setuptools is going to be already imported), 
because it has side effects.

Moreover, from a quick reading, `EnvironmentInfo` apparently gives you 
the latest MSVC version installed, not necessarily the one distutils is 
using for the build.

xoviat committed:
ENH: distutils: use the correct MSVC version

I attempted to ensure that the correct MSVC version was used and that when this eventually fails due to setuptools API changes, it will be simple to debug. Hopefully by then Python will have a stable API to get this information.

xoviat committed:
BUG: distutils: fix path comparison

VCTools returns list of strings, not string

xoviat committed:
BUG: distutils: fix string concatenation

xoviat committed:
BUG: distutils: fix case-insensitive paths

xoviat committed:
BUG: distutils: use normcase rather than .lower()

Norm case provides cross-platform path support.

xoviat commented:
@pv The code fails gracefully and correctly identifies the MSVC version:

```
  Trying to find MSVC version for: c:\program files (x86)\microsoft visual studio 14.0\vc\bin
  Trying version: 9.0
  Trying: c:\program files (x86)\microsoft visual studio 9.0\vc\vcpackages
  Trying: c:\program files (x86)\microsoft visual studio 9.0\vc\bin
  Trying version: 10.0
  Trying: c:\program files (x86)\microsoft visual studio 10.0\vc\vcpackages
  Trying: c:\program files (x86)\microsoft visual studio 10.0\vc\bin
  Trying version: 11.0
  Trying: c:\program files (x86)\microsoft visual studio 11.0\vc\vcpackages
  Trying: c:\program files (x86)\microsoft visual studio 11.0\vc\bin
  Trying version: 12.0
  Trying: c:\program files (x86)\microsoft visual studio 12.0\vc\vcpackages
  Trying: c:\program files (x86)\microsoft visual studio 12.0\vc\bin
  Trying version: 14.0
  Trying: c:\program files (x86)\microsoft visual studio 14.0\vc\vcpackages
  Trying: c:\program files (x86)\microsoft visual studio 14.0\vc\bin
  Found redist: C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\redist\x86\Microsoft.VC140.CRT
  copying C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\redist\x86\Microsoft.VC140.CRT\msvcp140.dll -> build\lib.win32-3.6\scipy\extra-dll\msvcp140.dll
```

It should also be robust enough to last for some years because setuptools should handle paths on newer visual studio versions. That's the idea, anyway.

pv commented:
Didn't find redist for py3.6-amd64 & py3.5-amd64: https://ci.appveyor.com/project/scipy/scipy-wheels/build/1.0.47/job/rmwcm4ontddcyh5c

xoviat committed:
BUG: distutils: try more locations for the directories

MSVC can compile X64 on an X86 machine, which distutils seems to prefer.

xoviat commented:
I have opened pypa/setuptools#1165 so that hopefully when this breaks setuptools will have a stable API available.

xoviat committed:
BUG: distutils: try even move locations

xoviat committed:
ENH: distutils: show contents of redist directory

xoviat commented:
@pv These issues have been resolved.

xoviat commented:
Actually hold off on merging this (though you can use it to build the
wheels). I am going to try to get setuptools to have a better API so that
the numpy code isn't so fragile.

pv commented:
Yeah, it looks a bit ugly, and might be better to wait for a better fix.

For scipy wheels (non-pip builds), in the meanwhile we can also do a 
simpler thing in appveyor
```
mkdir build\lib-win32-3.6\scipy\extra-dll
copy c:\program files (x86)\visual studio 
14.0\...\redist\...\msvcp140.dll build\lib-win32-3.6\scipy\extra-dll
```

xoviat committed:
ENH: distutils: cleanup and simplify the MSVC copy procedure

xoviat committed:
BUG: distutils: fix variable name

xoviat committed:
BUG: distutils: import stat module

xoviat committed:
BUG: distutils: fix warning

[skip-ci]

xoviat commented:
Unmarking as WIP because:
- DLLs are copied on Python 3.5 and 3.6 32 and 64 bit (tested)
- Code is not so fragile (although a private attribute is used, it is currently in Python 3.7 master)
- Warning message is displayed when DLLs are not copied
- Setuptools is not required

pv commented:
The next question is that this afaik is a copyright violation on some versions of MSVC (weird licenses forbid copying only one DLL)?

matthew-brett commented:
Maybe we could ask Steve Dower's advice about the copyright issues?

pv commented:
I think the license texts are more or less clear, but they vary between MSVC versions and the code doesn't check the version. So I think there should be whitelisting along the lines of
```
# Avoid distributing nonredistributable files
if 'debug_nonredist' in redist_folder.lower().split(os.path.sep):
    return _msvc_files_not_found()

redist_files = set(fn.lower for fn in os.listdir(redist_folder))
ok_sets = {
    frozenset(['msvcm90.dll', 'msvcp90.dll', 'msvcr90.dll', 'microsoft.vc90.crt.manifest']): 'all',
    frozenset(['msvcm140.dll', 'msvcp140.dll', 'msvcr140.dll']): 'any',
    # maybe more
}
set_status = ok_sets.get(redist_files)
if set_status == 'all':
    return redist_files
elif set_status == 'any':
    return [fn for fn in redist_files if fnmatch.fnmatch(pattern, fn)]
else:
    emit_msvc_warning
```

xoviat commented:
This only works on 3.5 and 3.6 so it shouldn't copy any DLL's from msvc 2008 or msvc 2010. However it should continue working in 3.7 and above.

xoviat commented:
It also doesn't copy any DLL's that should not be redistributed.

xoviat commented:
Just out of curiosity, is this PR moving forward? The discussion ended in a state where I don't think the PR could be merged. I personally think it should be, but there's no sense keeping the PR open if it's not going to be merged.

xoviat commented:
Guess that means no.
